Update gradle wrapper and build-tools version for flutter create.

asaarnak opened the pull request:
Motivation for this change was driven by:
1. Get the performance improvements of newer gradle(3.+), enabled daemon and incremental java compiler for example. For example: http://tools.android.com/tech-docs/new-build-system/2-5-alpha-gradle-plugin#TOC-What-to-expect-regarding-performance
2. The "Android Studio 2.3" "File->New Project" switched to newer gradle(3.3+) and "com.android.tools.build:gradle" to "2.3.0". 
3. And also many bugfixes: https://developer.android.com/studio/releases/gradle-plugin.html#revisions
Thanks.

Tested manually by "flutter create" with Idea "flutter run" and from android studio "run" android app.

asaarnak committed:
Update gradle wrapper and build-tools version.

Tested manually by "flutter create" with Idea "flutter run" and from android studio "run" android app.

asaarnak commented:
I signed it!

abarth commented:
@jakobr-google

asaarnak committed:
Update com.android.tools.build to 2.3.0

asaarnak commented:
Also updated the "packages/flutter_tools/gradle/flutter.gradle" dependency "com.android.tools.build" to version "2.3.0".

eseidelGoogle commented:
@jakobr-google @mit-mit

jakobr-google commented:
It does what it says on the tin, but why? Could you add the motivation for this change to the PR description, please?

mit-mit commented:
LGTM

Hixie commented:
This broke the build.

Hixie commented:
Specifically, basic_material_app__size is red since this checkin.

mit-mit commented:
I'll revert this until we better understand why that broke

asaarnak commented:
Does the test fail on server because of the apk release size is larger?

Running locally(ubuntu 16.04) the test seems ok, propably no benchmark history.
`flutter/dev/devicelab$ ../../bin/cache/dart-sdk/bin/dart bin/run.dart -t bin/tasks/basic_material_app__size.dart`
```
[basic_material_app__size] [STDOUT] Built build/app/outputs/apk/app-release.apk (7.9MB).
Task result:
{
  "success": true,
  "data": {
    "release_size_in_bytes": 8246558
  },
  "benchmarkScoreKeys": [
    "release_size_in_bytes"
  ]
}

••• Finished task "basic_material_app__size" •••
```

```
[✓] Flutter (on Linux, channel master)
    • Flutter at /home/allar/A/PROJECTS/flutter
    • Framework revision 578ca0a295 (2 hours ago), 2017-03-21 23:14:55
    • Engine revision 6b9eda4a40
    • Tools Dart version 1.23.0-dev.9.0

[✓] Host Executable Compatibility
    • Downloaded executables execute on host

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.2)
    • Android SDK at /home/allar/Android/Sdk
    • Platform android-25, build-tools 25.0.2
    • ANDROID_HOME = /home/allar/Android/Sdk
    • Java binary at: /usr/lib/jvm/java-8-oracle/bin/java
    • Java version: Java(TM) SE Runtime Environment (build 1.8.0_121-b13)

[✓] Android Studio (version 2.3)
    • Android Studio at /home/allar/A/PROGRAMS/IDE/android-studio
    • Gradle version 3.2

[✓] Gradle (3.4.1)
    • gradle-dir = /home/allar/A/PROGRAMS/GRADLE/latest/
    • Flutter supports building with Gradle from Android Studio.
      Consider removing your gradle-dir setting by running:
      flutter config --gradle-dir=
```

yjbanov commented:
Here's why it failed:

```
FAILURE: Build failed with an exception.

* Where:
Build file '/private/var/folders/_5/bqxnl8nn353dps4d6fzg77x00000gn/T/sample_flutter_app/android/app/build.gradle' line: 14

* What went wrong:
A problem occurred evaluating project ':app'.
> Failed to apply plugin [id 'com.android.application']
   > Minimum supported Gradle version is 3.3. Current version is 2.14.1. If using the gradle wrapper, try editing the distributionUrl in /private/var/folders/_5/bqxnl8nn353dps4d6fzg77x00000gn/T/sample_flutter_app/android/gradle/wrapper/gradle-wrapper.properties to gradle-3.3-all.zip
```

Should I upgrade Gradle in the devicelab or is this a regression?

yjbanov commented:
Notably, `flutter doctor` didn't complain:

```
[✓] Flutter is fully installed. (on Mac OS X 10.11.6 15G1217, channel unknown)
[✓] Host Executable Compatibility is fully installed.
[✓] Android toolchain - develop for Android devices is fully installed. (Android SDK 25.0.2)
[✓] iOS toolchain - develop for iOS devices is fully installed. (Xcode 7.3.1)
[✓] Android Studio is fully installed. (version 2.2)
[✓] Connected devices is fully installed.
```

asaarnak commented:
Problems with this PR:
- `flutter build apk --release` depends on `flutter config --gradle-dir=`(defaults to Gradle version shipped with Android Studio, currently too old so the tests failed)

- Android Plugin and Gradle version compatibility table: https://developer.android.com/studio/releases/gradle-plugin.html#updating-gradle
From the link above: "For the best performance, you should use the latest possible version of both Gradle and the Android plugin."

- 3.3+ Gradle required, but these defaults fail both:
`[✓] Android Studio is fully installed. (version 2.2)    • Gradle 2.14.1.`
`[✓] Android Studio (version 2.3)    • Gradle version 3.2`

- Android Studio itself generates gradle wrapper to support the `3.3+` if you create `New->Project`

Possible solution:
- `flutter create` template could also use gradle wrapper like Android Studio `New->Project`
- Bad: If user doesn't have the version of gradlew executed(cached) before, then ~70MB is downloaded.
- Good: Removes `--gradle-dir=` dependency from `flutter create`
- Good: Allows using newer Gradle and Android Plugin version combo in `flutter create template`

- Could create PR to futher discuss it. The test succeeded now.
Added these 4 files to `packages/flutter_tools/templates/create` 
and updated the .gitignore to not ignore them.
```
gradle/gradle-wrapper.jar
gradel/gradle-wrapper.properties
gradlew
gradlew.bat
```

jakobr-google commented:
We've debated having the gradle wrapper checked in before, but decided against it.

I think we'll want to hold off on retrying this one, until Android Studio fully supports Gradle 3.3 (as in, ships with it). You can still upgrade your own projects after creating them.

asaarnak commented:
Ok, thanks.
Doing this PR.
I have learned that `flutter build` is using gradle wrapper first and then `--gradle-dir=`
So for my many projects, older and newer. 
I will not depend on `--gradle-dir=` and can have dedicated gradle version per project. :)

jakobr-google commented:
That's true - if you've installed the gradle wrapper on your project, we will use that. Otherwise, we go hunting for Gradle, looking first at the --gradle-dir config, then Android Studio, and finally $PATH.
